### PR TITLE
DocumentDB Swagger Spec

### PR DESCRIPTION
**Pull request by dmakwana**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 
### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).
### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [X] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

---

This change is [Reviewable](https://reviewable.io/reviews/azure/azure-rest-api-specs/603)

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@dmakwana__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (dimakwan). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: There is no need to specify api-version in the path. Path just contains path parameters.
Query parameters are defined in the parameters section with the property `"in": "query"`.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: Why would you want to name the operation PatchTags? May be that is what the API allows today. In the future if the API supports patching more properties then this name will not be relevant. It would be a breaking change.
The name of the operation should be "DatabaseAccounts_Patch"

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: We recommend plural nouns for the first part in the operationId. DatabaseAccount**s**_

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: patchParameters

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: Create**s** ..... or update**s** ...

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: Patch**es** . ..

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: Delete**s**

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: General Comment: DatabaseAccount**s**_

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: Change**s** - just following the guidelines of the documentation team while reviewing operation descriptions. They recommend to start the operation description with a Verb in continuous present tense.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: It gets confusing.. The api in it's url and the operation name has failoverPriority and the name of the model that represents the body parameter is FailoverPolicy.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: empty description not allowed.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: no empty description.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: no empty description.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: no empty description.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: no empty description.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: This is a weird implementation of CheckNameAvailability. Usually services in Azure return a response with 200 status code. The response body has properties like { nameAvailable, reason, message }. Is it possible for the service to follow this pattern?

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: may be a constraint of `"minimum": 0`, if the service does not accept negative values.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: `"readOnly": true`

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: `"readOnly": true`

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: are id and locationName readOnly as well ?

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: missing description

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: missing description

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: missing description

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: "x-ms-parameter-location": "method",

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: `"version": "2015-04-08"` However the description says 2016-04-02. Please provide one value. We have a swagger spec per api-version.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/documentdb.json]: arm-documentdb/2015-04-08/**swagger/**documentdb.json. We missed a folder. Please take a look at other specs. They have a swagger folder inside the api-version folder.

**Review by amarzavery — CHANGES_REQUESTED** [review]

File: arm-documentdb/2015-04-08/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/documentdb.json]: I will keep note of this to review with the team. Thanks!

**Comment by dmakwana** [comment]
@amarzavery Unfortunately, had to do a force push, but I've addressed the comments.

File: arm-documentdb/2015-04-08/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/documentdb.json]: This value is to be deprecated so no more values are to be created in the future

**Comment by mimig1** [comment]
Content writer review complete!

**Comment by dmakwana** [comment]
@amarzavery Let me know if there are any fixes needed

**Comment by mimig1** [comment]
@tbombach & @amarzavery & @erickson-doug just a reminder that this is waiting to be reviewed/merged. Needs to be merged by 11/4.

**Comment by amarzavery** [comment]
This will be reviewed before EOD today.

Thanks,
Amar

**Comment by amarzavery** [comment]
@mimig1 - Please make sure that the description in the spec for all the artifacts are as per the guidelines from the documentation team. Please sync with @erickson-doug  about this before EOD today. This will make sure there are no multiple PRs after this PR is merged.

**Comment by mimig1** [comment]
@amarzavery - I am the writer that reviewed this and I believe it meets the guidelines. This is my first swagger file though. I checked it using a swagger previewer and I also used grammarly.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: By describing the negative response codes you are indicating to AutoRest that the generated client should not throw exceptions on 404 and 400. Is that the intended behavior?
If no then please document a "default" response and provide the description about different status codes over there.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: Does it make sense to provide an enum for a readOnly property? And would you really want to make this modelAsString: false? What if you decide to add a new value called "Cancelling"? Then that would be a breaking change. 
I would recommend removing the enum Array and the x-ms-enum extension.
However, if you wish to keep the enum array then make sure to make `"x-ms-enum".modelAsString: true`

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: We need to document what the specific response codes actually mean for this context. As long as the generated client can distinguish between a 200/404/400, I think we can go ahead with this approach.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: We don't have an enum for provisioningState over here. Whereas we have it for "Location.provisioningState". 
I would recommend not having an enum  anywhere in the spec specially for "provisioningState" as it does more harm than any good (w.r.t the property being readOnly (the iser cannot set it) and the values can potentially change if the service decides to add/remove/modify the states).

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: Since "properties" property is getting flattened and it has "consistencyPolicy" and "locations" as the required properties, they should also be required after getting flattened. Hence, please add "properties" as a required property as well.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: Why can you only update tags in the PATCH? PATCH too is one form of update. It seems weird that databaseAccount properties like location and consistencyPolicy can be updated via PUT (CreateOrUpdate) but cannot be updated via PATCH.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: This is marked as a long running operation. But it only returns a 200 response. This doesn't confirm to the standard long running operation pattern in Azure.
There should be a 201 or 202 over here.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: This is marked as a long running operation. But it only returns a 200 response. This doesn't confirm to the standard long running operation pattern in Azure.
There should be a 201 or 202 over here.

**Review by amarzavery — CHANGES_REQUESTED** [review]

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (johanste) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: @amarzavery, I believe the minimum requirements in ARM for what you have be be able to modify in a PATCH is tags. However, I completely agree that the service shouldn't strive for the minimum here - it should include all properties that are updateable.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (johanste) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: My assumption is that, given the lack of an expected 201, the service does not support upsert (further implied by the lack of otherwise required parameters for create from the body parameter). 

Still need an answer to if a 202 is supposed to be modeled here...

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (johanste) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: I would suggest a different name for the parameter with the assumption that at some point, we want the PATCH to support *all* mutatable properties. 

updateParameters comes to mind :)

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (johanste) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: Is this the correct description for OK? It sounds like it belongs to the 202, no?

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (johanste) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: Out of curiosity, why the discrepancy between the locations, writeLocations and readLocations between request body parameters and response body? The way things are modeled now (which may very well be correct), you will have a somewhat odd client SDK for strongly typed languages...

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: This was done as per the ARM standard mentioned in the Resource Provider API v2 documentation which mentioned that a 200/201 would be sufficient as long as the provisioningState property gets populated and updated. And this also works with the current portal.

For the time being, would it be better to remove the long-running-operation tag?

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: This was done as per the ARM standard mentioned in the Resource Provider API v2 documentation which mentioned that a 200/201 would be sufficient as long as the provisioningState property gets populated and updated. And this also works with the current portal.

For the time being, would it be better to remove the long-running-operation tag?

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: I will work on getting the RP updated to add more patcheable properties

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: I will work on getting this straightened out for the newer DocumentDB RP

**Comment by dmakwana** [comment]
@amarzavery I've addressed the comments

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: This is how the method signature looks like in csharp generated code
```
public static void CheckNameAvailability(this IDatabaseAccountsOperations operations, string accountName)
```
It makes no sense to the user. Whether the name is available or not and if it is not available then what is the reason.

This what the Async version of the method returns
```
string _responseContent = null;
            if ((int)_statusCode != 200 && (int)_statusCode != 404 && (int)_statusCode != 400)
            {
                var ex = new Microsoft.Rest.Azure.CloudException(string.Format("Operation returned an invalid status code '{0}'", _statusCode));
                if (_httpResponse.Content != null) {
                    _responseContent = await _httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
                }
                else {
                    _responseContent = string.Empty;
                }
                ex.Request = new Microsoft.Rest.HttpRequestMessageWrapper(_httpRequest, _requestContent);
                ex.Response = new Microsoft.Rest.HttpResponseMessageWrapper(_httpResponse, _responseContent);
                if (_httpResponse.Headers.Contains("x-ms-request-id"))
                {
                    ex.RequestId = _httpResponse.Headers.GetValues("x-ms-request-id").FirstOrDefault();
                }
                if (_shouldTrace)
                {
                    Microsoft.Rest.ServiceClientTracing.Error(_invocationId, ex);
                }
                _httpRequest.Dispose();
                if (_httpResponse != null)
                {
                    _httpResponse.Dispose();
                }
                throw ex;
            }
            // Create Result
            var _result = new Microsoft.Rest.Azure.AzureOperationResponse();
            _result.Request = _httpRequest;
            _result.Response = _httpResponse;
            if (_httpResponse.Headers.Contains("x-ms-request-id"))
            {
                _result.RequestId = _httpResponse.Headers.GetValues("x-ms-request-id").FirstOrDefault();
            }
            if (_shouldTrace)
            {
                Microsoft.Rest.ServiceClientTracing.Exit(_invocationId, _result);
            }
            return _result;
```

the body property of the AzureOperationResponse will be null.

Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: This is how the response of checkNameAvailability should look like: https://github.com/Azure/azure-rest-api-specs/blob/02ca4c505e064e8cfe646397433249f3965bd272/arm-storage/2016-01-01/swagger/storage.json#L434

This is the corresponding operation in storage https://github.com/Azure/azure-rest-api-specs/blob/02ca4c505e064e8cfe646397433249f3965bd272/arm-storage/2016-01-01/swagger/storage.json#L26

**Review by amarzavery — CHANGES_REQUESTED** [review]

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: For the time being, we need these response codes in the swagger-generated documentation. We will improve the response codes moving forward so it works well with the auto-generated client.

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: The check is done on the server side already. To avoid risking a mis-constructed pattern, I think it'll be okay to skip this for the time being.

**Comment by mimig1** [comment]
@dmakwana - Are we waiting on signoff from @amarzavery?

**Comment by dmakwana** [comment]
@mimig1 yes

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (hovsepm) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: >"nextLinkName": null [](start = 10, length = 20)

Are you sure there will be no more than one page of result for any subscription? if server side exposed ability to get several pages of results then you put developers in the situation that they will not be able to get second page easily.

Comment (dmakwana) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: Yes, we return all of them at once

File: arm-documentdb/2015-04-08/swagger/documentdb.json
Comment (amarzavery) [inline arm-documentdb/2015-04-08/swagger/documentdb.json]: DatabaseAccounts_CheckNameExists

**Comment by dmakwana** [comment]
@amarzavery ready

**Review by amarzavery — APPROVED** [review]

**Comment by amarzavery** [comment]
Will merge once the CI passes

**Pull request merged**
